### PR TITLE
glob-cpp: init at 0.0.1

### PR DESCRIPTION
**Pull request by ken-matsui**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

* related to https://github.com/NixOS/nixpkgs/pull/178258

### CODE REVIEW

**Comment by AndersonTorres** [comment]
Absolute failure. Most certainly because PackageProject explicitly invokes git and tries to download things.

**Comment by ken-matsui** [comment]
Yea, so awaiting https://github.com/NixOS/nixpkgs/pull/206489 to be merged: https://github.com/NixOS/nixpkgs/pull/178258#issuecomment-1356231196

File: pkgs/development/libraries/glob-cpp/default.nix [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    hash = "sha256-2y+a7YFBiYX8wbwCCWw1Cm+SFoXGB3ZxLPi/QdZhcdw=";
+  };
+
+  preConfigure = ''
+    mkdir -p ${placeholder "out"}/share/cpm
+    cp ${cpm-cmake}/share/cpm/CPM.cmake ${placeholder "out"}/share/cpm/CPM_0.27.2.cmake
+  '';
+
+  cmakeFlags = [
+    "-DCPM_SOURCE_CACHE=${placeholder "out"}/share"
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/glob-cpp/default.nix:26]:
```suggestion
    mkdir -p $out/share/cpm
    cp ${cpm-cmake}/share/cpm/CPM.cmake $out/share/cpm/CPM_0.27.2.cmake
  '';

  cmakeFlags = [
    "-DCPM_SOURCE_CACHE=$out/share"
```
can we avoid hardcoding the version number?

Comment (ken-matsui) [reply pkgs/development/libraries/glob-cpp/default.nix:26]: Unfortunately, we cannot. The version is hardcoded in the [source repository](https://github.com/p-ranav/glob/blob/v0.0.1/cmake/CPM.cmake).

**Comment by AndersonTorres** [comment]
Have you tried to build it locally? There are many build errors occurring because of "command not found"...

**Comment by ken-matsui** [comment]
@AndersonTorres 

Reverted the commit. This should work.

**Comment by ken-matsui** [comment]
@AndersonTorres

Could you please review this PR again?

**Comment by AndersonTorres** [comment]
According to the homepage, they have a small C++ program as example. Can you create a test/checkPhase from it?

**Comment by ken-matsui** [comment]
@AndersonTorres 

To test the standalone binary, we have to build this with `-Hall` to generate the binary. It will be an extra build step, but do you still want me to do so?

```console
$ cmake -Bbuild
$ cmake --build build
[ 50%] Building CXX object CMakeFiles/Glob.dir/source/glob.cpp.o
[100%] Linking CXX static library libglob.a
[100%] Built target Glob

$ cmake -Hall -Bbuild
$ cmake --build build
[ 25%] Building CXX object _deps/glob-build/CMakeFiles/Glob.dir/source/glob.cpp.o
[ 50%] Linking CXX static library libglob.a
[ 50%] Built target Glob
[ 75%] Building CXX object standalone/CMakeFiles/GlobStandalone.dir/source/main.cpp.o
[100%] Linking CXX executable glob
[100%] Built target GlobStandalone
```

**Comment by AndersonTorres** [comment]
I am not understanding. Is there any incompatibility when the standalone binary is built?

**Comment by ken-matsui** [comment]
There's no incompatibility. Some package source maintainers do not like extra build steps just for testing; that's why I asked.

**Comment by AndersonTorres** [comment]
This is not how we work at Nixpkgs. Tests are valuable. Indeed, many package sets (like Python libraries) are tested by default, being the `doCheck = false;` enabled and documented when failing.

There is no reason to be afraid of losing extra processor cycles for testing. This was already a weak argument in the times of single-core 32-bit Pentium - after all, a bad software running in such a slow machine is a waste of time and money; it is better to lose time testing that losing time in production setting.

**Comment by ken-matsui** [comment]
I understand. Let me create a test.

**Comment by ken-matsui** [comment]
* We cannot simply pass `-Hall` as the cmakeFlags
* The standalone binary requires an additional library; meaning that we have to do `fetchFromGitHub` it (https://github.com/p-ranav/glob/blob/b66732b0e72b3122e3552b6195ef3d64be08f35b/standalone/CMakeLists.txt#L13)

I don't have enough time to resolve these problems to accomplish adding tests. If anyone wants to do it just go ahead.

**Pull request closed without merging**